### PR TITLE
Move ThirdPartyEventRules into module_api/callbacks

### PR DESCRIPTION
**Pull request by anoadragon453**
A follow-on from https://github.com/matrix-org/synapse/pull/15453, in an effort to move all Synapse module callback classes to a central location, rather than spread out throughout the codebase.

* The first commit moves the `synapse.events.third_party_rules.ThirdPartyEventRules` module to `synapse.module_api.callbacks.third_party_event_rules.ThirdPartyEventRulesModuleApiCallbacks`, and updates the references in `ModuleApi`.
* The second commit updates all other points in the codebase that call out to third party rules modules for checking events/user actions. There's a lot of files changed in this commit, but the changes are pretty repetitive.

### CODE REVIEW

File: synapse/module_api/callbacks/__init__.py [hunk 1]
```diff
@@ -24,8 +24,11 @@
     SpamCheckerModuleApiCallbacks,
 )
 
+from .third_party_event_rules_callbacks import ThirdPartyEventRulesModuleApiCallbacks
```
Comment (clokep) [inline synapse/module_api/callbacks/__init__.py:27]: FWIW I very much prefer absolute imports...

File: synapse/handlers/federation.py [hunk 2]
```diff
@@ -169,7 +169,9 @@ def __init__(self, hs: "HomeServer"):
 
         self._room_backfill = Linearizer("room_backfill")
 
-        self.third_party_event_rules = hs.get_third_party_event_rules()
+        self._third_party_event_rules = (
+            hs.get_module_api_callbacks().third_party_event_rules
+        )
```
Comment (clokep) [inline synapse/handlers/federation.py:174]: Thanks for adding the underscores as we did this!

File: tests/server.py [hunk 3]
```diff
@@ -73,11 +73,13 @@
...
 from synapse.config.homeserver import HomeServerConfig
 from synapse.events.presence_router import load_legacy_presence_router
-from synapse.events.third_party_rules import load_legacy_third_party_event_rules
 from synapse.handlers.auth import load_legacy_password_auth_providers
 from synapse.http.site import SynapseRequest
 from synapse.logging.context import ContextResourceUsage
 from synapse.module_api.callbacks.spamchecker_callbacks import load_legacy_spam_checkers
+from synapse.module_api.callbacks.third_party_event_rules_callbacks import (
+    load_legacy_third_party_event_rules,
+)
```
Comment (clokep) [inline tests/server.py:82]: Any plans to remove legacy stuff? 👼

**Review by clokep — APPROVED** [review]
Seems OK if CI is happy.

File: tests/server.py (hunk 3 above)
Comment (anoadragon453) [reply tests/server.py:82]: Yes... I will get back to this. There's a few to eliminate:

https://github.com/matrix-org/synapse/blob/a924433e24e2ac81ca69665d8c737c5754b6c734/tests/server.py#L1053-L1056

File: synapse/module_api/callbacks/__init__.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/module_api/callbacks/__init__.py]: This is the third time I've forgotten to change that after cherry-picking :sob: Thanks for the catch.

Comment (clokep) [reply synapse/module_api/callbacks/__init__.py]: Annoyingly Pycharm's auto-add import seems to really prefer relative imports. 😢

**Pull request merged**